File: ergon-dashboard/src/lib/contracts/contextEvents.ts
// ergon-dashboard/src/lib/contracts/contextEvents.ts
/**
 * TypeScript types for run_context_events — mirrors Python ContextEventPayload.
 * Must stay in sync with ergon_core/ergon_core/core/persistence/context/event_payloads.py
 */

export type ContextEventType =
  | "system_prompt"
  | "user_message"
  | "assistant_text"
  | "tool_call"
  | "tool_result"
  | "thinking";

export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };

export interface TokenLogprob {
  [key: string]: JsonValue | undefined;
  token: string;
  logprob: number;
  top_logprobs?: Record<string, JsonValue>[];
}

export type ContextEventPayload =
  | { event_type: "system_prompt"; text: string }
  | { event_type: "user_message"; text: string; from_worker_key: string | null }
  | {
      event_type: "assistant_text";
      text: string;
      turn_id: string;
      turn_token_ids: number[] | null;
      turn_logprobs: TokenLogprob[] | null;
    }
  | {
      event_type: "tool_call";
      tool_call_id: string;
      tool_name: string;
      args: Record<string, unknown>;
      turn_id: string;
      turn_token_ids: number[] | null;
      turn_logprobs: TokenLogprob[] | null;
    }
  | {
      event_type: "tool_result";
      tool_call_id: string;
      tool_name: string;
      result: unknown;
      is_error: boolean;
    }
  | {
      event_type: "thinking";
      text: string;
      turn_id: string;
      turn_token_ids: number[] | null;
      turn_logprobs: TokenLogprob[] | null;
    };

export interface ContextEventState {
  id: string;
  runId: string;
  taskExecutionId: string;
  taskNodeId: string;
  workerBindingKey: string;
  sequence: number;
  eventType: ContextEventType;
  payload: ContextEventPayload;
  createdAt: string;
  startedAt: string | null;
  completedAt: string | null;
}

File: ergon_core/ergon_core/core/views/dashboard_events/context_events.py
"""Dashboard event projections for persisted context event rows."""

from typing import cast
from uuid import UUID

from ergon_core.core.persistence.context.models import SampleContextEvent
from ergon_core.core.shared.context_parts import ContextEventType
from ergon_core.core.views.dashboard_events.contracts import DashboardContextEventEvent


def context_event_to_dashboard_event(
    event: SampleContextEvent,
    execution_task_map: dict[UUID, UUID],
) -> DashboardContextEventEvent | None:
    task_id = execution_task_map.get(event.task_attempt_id)
    if task_id is None:
        return None
    return DashboardContextEventEvent(
        id=event.id,
        sample_id=event.sample_id,
        task_attempt_id=event.task_attempt_id,
        task_id=task_id,
        worker_binding_key=event.worker_binding_key,
        sequence=event.sequence,
        event_type=cast(ContextEventType, event.event_type),
        payload=event.parsed_payload(),
        created_at=event.created_at,
        started_at=event.started_at,
        completed_at=event.completed_at,
    )
